diff -up xinit-1.3.4/startx.cpp~ xinit-1.3.4/startx.cpp
--- xinit-1.3.4/startx.cpp~	2015-03-18 12:13:11.000000000 +0100
+++ xinit-1.3.4/startx.cpp	2015-03-18 12:49:49.445624223 +0100
@@ -199,6 +199,8 @@ if [ x"$server" = x ]; then
     if expr match "$tty" '^/dev/tty[0-9]\+$' > /dev/null; then
         tty_num=$(echo "$tty" | grep -oE '[0-9]+$')
         vtarg="vt$tty_num -keeptty"
+        XCOMM Fedora specific mod to make X run as non root
+        export XORG_RUN_AS_USER_OK=1
     fi
 #endif
 
